discv5 dns bootstrap

kaiserd opened the pull request:
## Background

This PR addresses #880.
The current draft implementation passed a test on an experimental fleet.
This fleet server can be used for testing:
`enrtree://AKRUVB7WRORHGL5IIHKHEXGALD5IWWAL7UIVOUZ2FKE4ORV5WOX76@waku.337.ovh`
(I will keep this running during the testing phase; might be down sometimes during update phases.)

The following explains my implementation choices.
(No strong opinion on these, and happy to change if desired :))


### Introducing a separate setup phase for bootstrap nodes

Bootstrap nodes are passed to the discv5 object during the `initialize node` phase while they are dialed by Waku relay in the `start node` phase.
To avoid retrieving bootstrap nodes twice, having a separate phase for retrieving them before `initialize node` seemed like a solution with not too much restructuring.
Also, the new `retrieve dynamic bootstrap nodes` phase can be extended to retrieve dynamic bootstrap nodes from other sources, too (e.g. mDNS).

### ENR in RemotePeerInfo

This PR extends `RemotePeerInfo` storing ENRs of discovered nodes.
[DNS discovery](https://rfc.vac.dev/spec/10/#discovery-methods) retrieves ENRs anyway. While storing the ENR was not useful before because Waku only uses multiaddrs, it is now useful with the introduction of discv5, which uses the ENR format.
Storing the ENRs also avoids additional conversions.

(I am not sure if `Option` is the idiomatic Nim way for introducing an optional field. I did not find information on this in the [style guide](https://status-im.github.io/nim-style-guide/00_introduction.html). Options feel a bit clunky in Nim.)

## Todo

If the current implementation choices are fine, I would address the following

* [x] remove double conversion of ENRs from string to Record format and vice versa
* ~extend local peerInfo with ENR~ `peerInfo` is in libp2p
   - a future PR could generate an ENR [here](https://github.com/status-im/nim-waku/blob/b8da352a56b49642f198f30c45946669e055c0c1/waku/v2/utils/peers.nim#L160)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 6a5351b6 | [#1](https://ci.status.im/job/nim-waku/job/prs/job/macos/job/PR-888/1/) | 2022-03-10 17:29:58 | ~19 min | `macos` | [:package:`bin`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/wakunode2-220310-171015-6a5351-macos.bin) |

jm-clius commented on waku/v2/node/wakunode2.nim:
```diff
@@ -1257,44 +1265,53 @@ when isMainModule:
     (pStorage, mStorage) = setupStorageRes.get()
 
-  debug "2/6 Initializing node"
+  debug "2/7 Retrieve dynamic boostrap nodes"
```
```suggestion
  debug "2/7 Retrieve dynamic bootstrap nodes"
```
:)

jm-clius commented on waku/v2/node/wakunode2.nim:
```diff
@@ -1257,44 +1265,53 @@ when isMainModule:
+  var dynamicBootstrapNodes: seq[RemotePeerInfo]
+  let dynamicBootstrapNodesRes = retrieveBootstrapNodes(conf)
+  if dynamicBootstrapNodesRes.isErr:
+    error "2/7 Retrieving dynamic boostrap nodes failed. Continuing without dynamic boostrap nodes."
```
```suggestion
    error "2/7 Retrieving dynamic bootstrap nodes failed. Continuing without dynamic boostrap nodes."
```
:)

jm-clius commented:
This approach makes sense to me! Feel free to mark as ready for review once it is. :)

kaiserd commented:
I now opened the issue for review after addressing the double ENR conversion.
The tests on my experimental fleet still run successfully.

I added a second constructor for WakuDiscoveryV5, which takes `seq[enr.Record]` instead of a `seq[string]` (EnrUri).
This saves conversations. (The old constructor is not used anymore; but it might still make sense to keep it for potential future use.)

Imo, generating an ENR in [`toRemotePeerInfo`](https://github.com/status-im/nim-waku/blob/b8da352a56b49642f198f30c45946669e055c0c1/waku/v2/utils/peers.nim#L160) could be addressed in a separate Issue/PR. Let me know, if I should do this in this PR. Currently, this generated ENR would not be used.


[Here](https://github.com/status-im/nim-waku/blob/c838b487ab02359119991dfcee83c8a8a783738a/waku/v2/node/wakunode2.nim#L1104), I tried to do something like

```nim
var discv5BootstrapEnrs = dynamicBootstrapNodes
                            .filterIt(it.enr.isSome() and
                              it.enr.get().toTypedRecord().tryGet().udp.isSome()) # TODO: idiomatic error handling. 
                            .mapIt(it.enr.get())
```

However, filter and map are not composable without copying, and I could not find a way of addressing the potential error on tryGet() inline.
Does Nim provide something like result combinators in Rust?

I also found collect, which would allow doing something along the lines of:

```nim
let dynamicBootstrapEnrs = collect(newSeq):
                            for n in dynamicBootstrapNodes:
                              if n.enr.isSome():
                                n.enr.get()
```

However, this requires `std/sugar`, uses a macro, and is not much more concise compared to a regular for loop (in this example, I I did not do the udp port check, which is why it looks more concise.)

kaiserd commented:
Because the `WakuDiscoveryV5` constructores only differ in the type of the ENR seq, passing an empty seq without type does not work anymore. The type has to be specified like  [here](https://github.com/status-im/nim-waku/blob/f15990d93149ea90d1f397f08332c8e3495ddf60/tests/v2/test_waku_discv5.nim#L55).

oskarth commented on waku/v2/node/wakunode2.nim:
```diff
@@ -1013,9 +1013,34 @@ when isMainModule:
 
-  # 2/6 Initialize node
+  # 2/7 Retrieve dynamic bootstrap nodes
+  proc retrieveDynamicBootstrapNodes(conf: WakuNodeConf): Result[seq[RemotePeerInfo], cstring] =
```
Is there a reason we are using `cstring` here as opposed to hiding this in a lower level API? Seems like an implementation detail that'd be better to abstract away to either a string or some other relevant type. (I realize this might be a consequence of lower level API)

kaiserd replied:
There is no reason. I converted the Result to `SetupResult`, which all other setup methods use as well. This is much clearer.  Thank you :).

oskarth commented on waku/v2/node/wakunode2.nim:
```diff
@@ -1013,9 +1013,34 @@ when isMainModule:
 
-  # 2/6 Initialize node
+  # 2/7 Retrieve dynamic bootstrap nodes
+  proc retrieveDynamicBootstrapNodes(conf: WakuNodeConf): Result[seq[RemotePeerInfo], cstring] =
```
Does this block? One concern would be if the node gets busy discovering nodes and isn't able to start functioning earlier

kaiserd replied:
Yes it blocks. Agreed.
The blocking stems form [waku_dnsdisc](https://github.com/status-im/nim-waku/blob/e6fdbdb42ac8592c02811d4a2709d4d17e065d0c/waku/v2/node/dnsdisc/waku_dnsdisc.nim#L52).
Imo, we should address this in a separate issue.

For now, this should not be too severe, because it is "just" blocking DNS queries. These DNS records should be available in the next setup step and would be waited for anyways.
I think the DNS disc API should be async, but the setup functions should be sync because the steps sequentially depend on each other.

Should I open a new issue or take further steps within this PR?

oskarth replied:
Agree and opening a new issue sounds good, thanks!

oskarth commented on waku/v2/node/wakunode2.nim:
```diff
@@ -1075,11 +1100,25 @@ when isMainModule:
         discoveryConfig = DiscoveryConfig.init(
           conf.discv5TableIpLimit, conf.discv5BucketIpLimit, conf.discv5BitsPerHop)
 
+      # select dynamic bootstrap nodes that have an ENR containing a udp port
```
Maybe brief comment mentioning why this is necessary, i.e. Discovery v5 is only UDP, or link to spec

oskarth commented on waku/v2/utils/peers.nim:
```diff
@@ -145,11 +150,12 @@ proc toRemotePeerInfo*(enr: enr.Record): Result[RemotePeerInfo, cstring] =
 proc toRemotePeerInfo*(peerInfo: PeerInfo): RemotePeerInfo =
   RemotePeerInfo.init(peerInfo.peerId,
                       peerInfo.addrs,
+                      none(enr.Record), # TODO
```
What's the TODO referring to here?

kaiserd replied:
I replaced the TODO with a comment.

oskarth reviewed (commented):
Nice one! Some minor comments

kaiserd commented:
I addressed the feedback and also [allowed a `udp6` port in the ENR](https://github.com/status-im/nim-waku/blob/305735b250802810a05b758d8a521ab04a952e37/waku/v2/node/wakunode2.nim#L1112), because the [ENR spec](https://eips.ethereum.org/EIPS/eip-778) does not exclude ENRs that only have `udp6` but no `udp`.

jm-clius reviewed (approved):
LGTM! Thanks.
